native/membrane_videocompositor_opengl_rust/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+  rustler::resource!(State, env);
+  true
+}
+
+#[rustler::nif]
+/// Initialize the compositor. This function is intended to only be called by Elixir.
+fn init(first_video: RawVideo, second_video: RawVideo, out_video: RawVideo) -> Result<(rustler::Atom, ResourceArc<State>), rustler::Error> {
+  use crate::scene::Scene;
+  use crate::shaders::ShaderProgram;
+
+  if first_video.pixel_format != atoms::I420() || second_video.pixel_format != atoms::I420() || out_video.pixel_format != atoms::I420() {
+    return Err(rustler::Error::Term(Box::new(atoms::unsupported_pixel_format())));
+  }
+
+  let egl = &egl::API;
+
+  let display = egl
+    .get_display(egl::DEFAULT_DISPLAY)
+    .ok_or(rustler::Error::Atom("cant_get_default_display"))?;
+
+  egl.initialize(display).nif_err()?;
+
+  #[rustfmt::skip]
+  let attributes = [
+    egl::SURFACE_TYPE,  egl::PBUFFER_BIT,
+    egl::RED_SIZE,      8,
+    egl::GREEN_SIZE,    8,
+    egl::BLUE_SIZE,     8,
+    egl::CONFORMANT,    egl::OPENGL_ES3_BIT,
+    egl::NONE
+  ];
+
+  let config = egl
+    .choose_first_config(display, &attributes)
+    .nif_err()?
+    .expect("Got no compatible config");
+
+  #[rustfmt::skip]
+  let attributes = [
+    egl::CONTEXT_MAJOR_VERSION, 3,
+    egl::CONTEXT_MINOR_VERSION, 0,
+    egl::NONE,
+  ];
+
+  let context = egl
+    .create_context(display, config, None, &attributes)
+    .nif_err()?;
+
+  egl.make_current(display, None, None, Some(context)).nif_err()?;
+
+  gl::load(|name| {
+    egl
+      .get_proc_address(name)
+      .expect("Can't find a GLES procedure") as *const std::ffi::c_void
+  });
+
+  unsafe { gl::ClearColor(0.0, 0.0, 0.0, 1.0) }
+
+  let vertex_shader_code = include_str!("shaders/vertex.glsl");
+  let fragment_shader_code = include_str!("shaders/fragment.glsl");
+
+  let shader_program = ShaderProgram::new(vertex_shader_code, fragment_shader_code)?;
+  let mut scene = Scene::new(out_video.width, out_video.height, shader_program);
+  scene.add_video(
+    scene::VideoPlacementTemplate {
+      top_right: (1.0, 1.0),
+      top_left: (-1.0, 1.0),
+      bot_left: (-1.0, 0.0),
+      bot_right: (1.0, 0.0),
+      z_value: 0.0
+    },
+    first_video.width, second_video.height
+  );
+
+  scene.add_video(
+    scene::VideoPlacementTemplate {
+      top_right: (1.0, 0.0),
+      top_left: (-1.0, 0.0),
+      bot_left: (-1.0, -1.0),
+      bot_right: (1.0, -1.0),
+      z_value: 0.0
+    },
+    first_video.width, second_video.height
+  );
+
+
+
+  egl.make_current(display, None, None, None).nif_err()?;
+  Ok((atoms::ok(), ResourceArc::new(State::new(display, context, scene))))
+}
+
+#[rustler::nif]
+/// Join two frames passed as [binaries](rustler::Binary). This function is intended to only be called by Elixir.
+fn join_frames<'a>(env: rustler::Env<'a>, state: rustler::ResourceArc<State>, upper: rustler::Binary, lower: rustler::Binary) -> Result<(rustler::Atom, rustler::Term<'a>), rustler::Error> {
+  let ctx: BoundContext = state.bind_context().nif_err()?;
```

Comment:
type annotation can be elided

```suggestion
  let ctx = state.bind_context().nif_err()?;
```